### PR TITLE
Add documentation to SubmitSignedTransaction and actually make it work

### PR DESCRIPTION
**Pull request by tomusdrw**
With recent changes to crypto stuff (notably `IdentifyAccount` that makes the `AccountId` diverge from `PublicKey`) the `SubmitSignedTransaction` trait family was not really working correctly.

The main complications arise because of application-crypto type wrappers (`app-crypto`) and signed aggregators (`MultiSigner`), this requires a bit of `From/TryInto/Into` magic and some additional trait bounds, that on the first sight look unnecessary in this code :)

This PR is an attempt to fix that and also make it more convenient to use local keys for transaction signing (see `SigningAccountFinder`). This additional trait bound can be used to create and submit signed transactions from offchain calls using all local keys (intersected with some list of accounts).

Still WiP, as I want to work on tests a bit more, but wanted to make this visible for @kianenigma @gnunicorn and @andresilva

### CODE REVIEW

File: frame/system/src/offchain.rs [hunk 1]
```diff
@@ -130,9 +162,70 @@ pub trait SubmitUnsignedTransaction<T: crate::Trait, Call> {
...
+	///
+	/// This method should intersect given list of accounts with the ones
+	/// supported locally and submit signed transaction containing given `Call`
+	/// with every of them.
+	///
+	/// Returns a vector of results and account ids that were supported.
+	fn submit_signed(
+		accounts: impl IntoIterator<Item= T::AccountId>,
+		call: impl Into<Call> + Clone,
+	) -> Vec<(T::AccountId, Result<(), ()>)> {
```
Comment (andresilva) [inline frame/system/src/offchain.rs]: Do you think it makes sense to add an option to stop on the first successful submission?

Comment (tomusdrw) [reply frame/system/src/offchain.rs]: Why would that be useful? I think we should always submit with all accounts that are there (i.e. we do (or planned to) support mutliple validators on one node, right).

Comment (andresilva) [reply frame/system/src/offchain.rs]: In the equivocation report I was bailing out after the first success, but yeah probably makes sense to report from all available accounts anyway.

**Comment by tomusdrw** [comment]
Finally ready for review. Added tests and a bunch of utility methods (like `can_sign` - CC @gnunicorn )

**Comment by gavofyork** [comment]
tests failing?

**Comment by tomusdrw** [comment]
@gavofyork I believe it was caused by a new compiler version. Merged with master and it seems fine. Polkadot failure seems unrelated (some `clap` issues).

**Comment by kianenigma** [comment]
https://github.com/paritytech/substrate/pull/4517 is _eagerly_ waiting for this :D

**Comment by gavofyork** [comment]
@tomusdrw is this merge correct?

**Comment by tomusdrw** [comment]
Yup, thanks! Could I get a second review on this? @bkchr maybe if you have some time? :smiling_imp:

File: frame/system/src/offchain.rs [hunk 2]
```diff
@@ -130,9 +167,114 @@ pub trait SubmitUnsignedTransaction<T: crate::Trait, Call> {
...
+	/// Create and submit signed transactions from all local accounts.
+	///
+	/// This method submits a signed transaction from all local accounts
+	/// for given application crypto.
+	///
+	/// Returns a vector of results and account ids that were supported.
+	#[must_use]
+	fn submit_signed(
+		call: impl Into<Call> + Clone,
+	) -> Vec<(T::AccountId, Result<(), ()>)> {
```
[the remark this answers is not in the document]
Comment (tomusdrw) [reply frame/system/src/offchain.rs]: I like `impl` syntax on input parameters, but returning an `Iterator` seems a bit unnecessary for me. Also I don't think if it's even possible on a trait method.

File: bin/node/executor/tests/basic.rs [hunk 3]
```diff
@@ -0,0 +1,837 @@
...
+	System, TransactionPayment, Event,
+	TransferFee, TransactionBaseFee, TransactionByteFee,
+	constants::currency::*,
+};
+use node_primitives::{Balance, Hash};
+use wabt;
+use node_testing::keyring::*;
+
+mod common;
+use self::common::{*, sign};
```
[the remark this answers is not in the document]
Comment (tomusdrw) [reply bin/node/executor/tests/basic.rs]: This doesn't work as there is a conflict with glob imports (`common` and `keystore`), so we need to define `sign` explicitly.

File: frame/system/src/lib.rs [hunk 4]
```diff
@@ -628,11 +649,15 @@ impl<T: Trait> Module<T> {
 	}
 
 	/// Start the execution of a particular block.
+	///
+	/// Setting `is_for_inspection` flag will not cause the events
```
Comment (bkchr) [inline frame/system/src/lib.rs]: I think the comment can be removed? Or at least be fixed.

Comment (tomusdrw) [reply frame/system/src/lib.rs]: Good catch, thanks!

**Comment by tomusdrw** [comment]
@bkchr convinced me to change `offchain_worker` `RuntimeApi` and pass the entire header to make the initialization properly, I'm working on this right now, hence the `inprogress` tag.

**Pull request merged**